### PR TITLE
Feat: Ability to deploy requirements on Devnet

### PR DESCRIPTION
**Pull request by lgalabru**
#333 paved some paths for addressing a few problems, and we'll start addressing the pain point explained here https://github.com/hirosystems/clarinet/issues/346.
Developers want to build protocols on top of each other (composition), and they also want to test these integrated composition of protocols on their own Devnet.
In the flow proposed in this PR, when developers want to interact with a contract deployed on **Mainnet** (other protocol, or even a trait), they list it with the command:
```
clarinet requirements add SP3FBR2AGK5H9QBDH3EEN6DF8EK8JY7RX8QJ5SVTE.sip-010-trait-ft-standard 
```
which essentially update the `Clarinet.toml` manifest:

```toml
[[project.requirements]]
contract_id = "SP3FBR2AGK5H9QBDH3EEN6DF8EK8JY7RX8QJ5SVTE.sip-010-trait-ft-standard"
```

From there, the contract will be downloaded and cached locally (path customizable via the setting `cache_dir` in `Clarinet.toml`). The contract can be used in contracts (`contract-call?`, `use-trait`, etc), like any other local contract, using the mainnet address.
When writing tests, developers are writing their assertion, still using the `mainnet` address - in short, the source of truth always use the `mainnet` address.
When developers want to deploy to Devnet, Testnet, all the requirements and their references are automatically remapped and deployed.
By default, the deployer account is being used, but this can be customized with the following approach:

```bash
$ clarinet deployments generate --devnet
// Generate deployments/default.devnet-plan.yaml

$ cat deployments/default.devnet-plan.yaml
```

```yaml
---
id: 0
name: Devnet deployment
network: devnet
node: "http://localhost:20443"
plan:
  batches:
    - id: 0
      transactions:
        - requirement-publish:
            contract-id: SP3FBR2AGK5H9QBDH3EEN6DF8EK8JY7RX8QJ5SVTE.sip-010-trait-ft-standard
            remap-sender: ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM # <- Change address here
            path: "./.requirements/SP3FBR2AGK5H9QBDH3EEN6DF8EK8JY7RX8QJ5SVTE.sip-010-trait-ft-standard.clar"
            cost: 8400
        - contract-publish:
            contract-name: counter
            expected-sender: ST1PQHQKV0RJXZFY1DGX8MNSNYVE3VGZJSRTPGZGM
            path: contracts/counter.clar
            cost: 7010
```

There are several protections in place to make sure that a remap never happens on a mainnet deployment - a deployment plan targeting `mainnet` including a `requirement-publish` fails parsing.

This PR also address https://github.com/hirosystems/clarinet/issues/359, so developers should be able to rely exclusively on deployment plans when deploying their contracts even when they need to mix `contract-publish` and `contract-call` transactions for setting up their protocols.

### CODE REVIEW

File: src/types/mod.rs [hunk 1]
```diff
@@ -80,6 +80,20 @@ impl StacksNetwork {
             _ => false,
         }
     }
+
+    pub fn either_devnet_or_tesnet(&self) -> bool {
```
Comment (obycode) [inline src/types/mod.rs:84]:
```suggestion
    pub fn either_devnet_or_testnet(&self) -> bool {
```

File: src/types/mod.rs [hunk 2]
```diff
@@ -80,6 +80,20 @@ impl StacksNetwork {
...
+        match self {
+            StacksNetwork::Simnet | StacksNetwork::Devnet => true,
```
Comment (obycode) [inline src/types/mod.rs:86]: Shouldn't this be `Testnet`, not `Simnet`? Or should the function be renamed?

**Review by obycode — CHANGES_REQUESTED** [review]
Looks great @lgalabru! Just a few comments.

File: src/deployment/mod.rs [hunk 3]
```diff
@@ -236,12 +280,14 @@ pub fn update_session_with_contracts_executions(
     code_coverage_enabled: bool,
 ) -> BTreeMap<QualifiedContractIdentifier, Result<ExecutionResult, Vec<Diagnostic>>> {
     let mut results = BTreeMap::new();
+    // let mut remap_to_perform = vec![];
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/deployment/mod.rs]: Addressed in https://github.com/hirosystems/clarinet/pull/388/commits/895ba873a7a36f7218ce35e4f9e081fb5d4d18fb

File: src/deployment/mod.rs [hunk 4]
```diff
@@ -294,7 +340,8 @@ pub fn update_session_with_contracts_analyses(
     for batch in deployment.plan.batches.iter() {
         for transaction in batch.transactions.iter() {
             match transaction {
-                TransactionSpecification::ContractCall(_)
+                TransactionSpecification::RequirementPublish(_)
+                | TransactionSpecification::ContractCall(_)
                 | TransactionSpecification::ContractPublish(_) => {
                     panic!("emulated-contract-call and contract-publish are the only operations admitted in simnet deployments")
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/deployment/mod.rs]: Addressed in https://github.com/hirosystems/clarinet/pull/388/commits/895ba873a7a36f7218ce35e4f9e081fb5d4d18fb

File: src/deployment/types.rs [hunk 5]
```diff
@@ -447,6 +520,14 @@ impl DeploymentSpecification {
                                 TransactionSpecificationFile::ContractCall(spec) => {
                                     TransactionSpecification::ContractCall(ContractCallSpecification::from_specifications(spec)?)
                                 }
+                                TransactionSpecificationFile::RequirementPublish(spec) => {
+                                    if network.is_mainnet() {
+                                        return Err(format!("{} only supports transactions of type 'contract-call' and 'contract-publish", specs.network.to_lowercase()))
+                                    }
+                                    let spec = RequirementPublishSpecification::from_specifications(spec, base_path)?;
+                                    // contracts.insert(contract_id, (spec.source.clone(), spec.relative_path.clone()));
```
[the remark this answers is not in the document]
Comment (lgalabru) [reply src/deployment/types.rs]: I initially added this line in my initial draft, but with the approach implemented, we do not - removed in https://github.com/hirosystems/clarinet/pull/388/commits/895ba873a7a36f7218ce35e4f9e081fb5d4d18fb.

File: src/types/mod.rs (hunk 1 above)
Comment (lgalabru) [reply src/types/mod.rs]: Addressed in https://github.com/hirosystems/clarinet/pull/388/commits/895ba873a7a36f7218ce35e4f9e081fb5d4d18fb

File: src/types/mod.rs (hunk 2 above)
Comment (lgalabru) [reply src/types/mod.rs]: Addressed in https://github.com/hirosystems/clarinet/pull/388/commits/895ba873a7a36f7218ce35e4f9e081fb5d4d18fb

**Review by obycode — APPROVED** [review]
Looks great! Thanks @lgalabru!

**Pull request merged**